// @finalytic/utils/build/module/validator/numeric.js
import BaseSchema from './base';
export default class NumericSchema extends BaseSchema {
    /**
     * It represents  an (inclusive | exclusive) lower limit for a numeric instance.
     *
     * @param {number} minimum
     * @param {boolean} exclusive
     * @reference https://json-schema.org/latest/json-schema-validation.html#rfc.section.6.2.4
     * @returns {NumericSchema}
     */
    minimum(minimum, exclusive = false) {
        return this.copyWith({
            plain: exclusive ? { exclusiveMinimum: minimum } : { minimum },
        });
    }
    /**
     * It represents  an (inclusive | exclusive) upper limit for a numeric instance.
     *
     * @reference https://json-schema.org/latest/json-schema-validation.html#rfc.section.6.2.2
     * @returns {NumericSchema}
     */
    maximum(maximum, exclusive = false) {
        return this.copyWith({
            plain: exclusive ? { exclusiveMaximum: maximum } : { maximum },
        });
    }
    /**
     * It must be multiple of.
     *
     * @param {number} multipleOf
     * @reference https://json-schema.org/latest/json-schema-validation.html#rfc.section.6.2.1
     * @returns {NumericSchema}
     */
    multipleOf(multipleOf) {
        return this.copyWith({ plain: { multipleOf } });
    }
    /**
     * Make schema optional in {ObjectSchema}
     * @returns {NumericSchema}
     */
    optional() {
        return this.copyWith({ isRequired: false });
    }
}
