// @finalytic/utils/build/module/utils/finance/currency.js
import formatter from 'currency-formatter';
const { format } = formatter;
export function getLocale() {
    const env = process.env;
    const language = env.LANG || env.LANGUAGE || env.LC_ALL || env.LC_MESSAGES;
    return language?.replace('_', '-').split('.')[0] || 'en-GB';
}
export function formatCurrency(a, currency, locale, precision = 2) {
    if (!currency) {
        currency = 'GBP';
    }
    if (!locale) {
        locale = 'en-GB';
    }
    if (!a) {
        a = 0;
    }
    const amount = typeof a === 'string' ? Number.parseFloat(a) : a;
    currency = currency.toUpperCase();
    let value = format(amount, {
        code: currency,
        locale: locale,
        precision,
    }).replace(/\s/g, ' ');
    if (currency?.toLowerCase() === 'mxn') {
        value = value.replace('$', 'mex$');
    }
    return value;
}
export function formatCurrencyNumber(a, locale) {
    const options = {
        minimumFractionDigits: 2,
        maximumFractionDigits: 2,
    };
    return Number(a).toLocaleString(locale, options);
}
