// @finalytic/utils/build/module/fetch-to-curl.js
export const generateMethod = (options) => {
    const method = options.method;
    if (!method)
        return '';
    const type = {
        GET: ' -X GET',
        POST: ' -X POST',
        PUT: ' -X PUT',
        PATCH: ' -X PATCH',
        DELETE: ' -X DELETE',
        HEAD: ' -X HEAD',
        OPTIONS: ' -X OPTIONS',
    };
    return type[method.toUpperCase()] || '';
};
export const isInstanceOfHeaders = (val) => {
    if (typeof Headers !== 'function') {
        return false;
    }
    return val instanceof Headers;
};
const getHeaderString = (name, val) => ` -H "${name}: ${`${val}`.replace(/(\\|")/g, '\\$1')}"`;
export const generateHeader = (options = {}) => {
    const { headers } = options;
    let isEncode = false;
    let headerParam = '';
    if (isInstanceOfHeaders(headers)) {
        for (const [name, val] of headers) {
            if (name.toLocaleLowerCase() !== 'content-length') {
                headerParam += getHeaderString(name, val);
            }
            if (name.toLocaleLowerCase() === 'accept-encoding') {
                isEncode = true;
            }
        }
    }
    else if (headers) {
        for (const [name, val] of Object.entries(headers)) {
            if (name.toLocaleLowerCase() !== 'content-length') {
                headerParam += getHeaderString(name, String(val));
            }
            if (name.toLocaleLowerCase() === 'accept-encoding') {
                isEncode = true;
            }
        }
    }
    return {
        params: headerParam,
        isEncode,
    };
};
export function escapeBody(body) {
    if (typeof body !== 'string')
        return body;
    return body.replace(/'/g, `'\\''`);
}
export function generateBody(body) {
    if (!body)
        return '';
    if (typeof body === 'object') {
        return ` --data-binary '${escapeBody(JSON.stringify(body))}'`;
    }
    return ` --data-binary '${escapeBody(body)}'`;
}
export function generateCompress(isEncode) {
    return isEncode ? ' --compressed' : '';
}
export const fetchToCurl = (requestInfo, requestInit) => {
    let url;
    let options;
    /**
     * initialization with an empty object is done here to
     * keep everything backwards compatible to 0.4.0 and below
     */
    if (typeof requestInfo === 'string' || requestInfo instanceof URL) {
        url = requestInfo;
        options = requestInit || {};
    }
    else {
        url = requestInfo?.url;
        options = requestInfo || {};
    }
    const { body } = options;
    const headers = generateHeader(options);
    return `curl '${url}'${generateMethod(options)}${headers.params || ''}${generateBody(body)}${generateCompress(headers.isEncode)}`;
};
