// @finalytic/utils/build/main/validator/numeric.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_1 = __importDefault(require("./base"));
class NumericSchema extends base_1.default {
    /**
     * It represents  an (inclusive | exclusive) lower limit for a numeric instance.
     *
     * @param {number} minimum
     * @param {boolean} exclusive
     * @reference https://json-schema.org/latest/json-schema-validation.html#rfc.section.6.2.4
     * @returns {NumericSchema}
     */
    minimum(minimum, exclusive = false) {
        return this.copyWith({
            plain: exclusive ? { exclusiveMinimum: minimum } : { minimum },
        });
    }
    /**
     * It represents  an (inclusive | exclusive) upper limit for a numeric instance.
     *
     * @reference https://json-schema.org/latest/json-schema-validation.html#rfc.section.6.2.2
     * @returns {NumericSchema}
     */
    maximum(maximum, exclusive = false) {
        return this.copyWith({
            plain: exclusive ? { exclusiveMaximum: maximum } : { maximum },
        });
    }
    /**
     * It must be multiple of.
     *
     * @param {number} multipleOf
     * @reference https://json-schema.org/latest/json-schema-validation.html#rfc.section.6.2.1
     * @returns {NumericSchema}
     */
    multipleOf(multipleOf) {
        return this.copyWith({ plain: { multipleOf } });
    }
    /**
     * Make schema optional in {ObjectSchema}
     * @returns {NumericSchema}
     */
    optional() {
        return this.copyWith({ isRequired: false });
    }
}
exports.default = NumericSchema;
