// @finalytic/utils/build/main/utils/finance/currency.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatCurrencyNumber = exports.formatCurrency = exports.getLocale = void 0;
const currency_formatter_1 = __importDefault(require("currency-formatter"));
const { format } = currency_formatter_1.default;
function getLocale() {
    const env = process.env;
    const language = env.LANG || env.LANGUAGE || env.LC_ALL || env.LC_MESSAGES;
    return language?.replace('_', '-').split('.')[0] || 'en-GB';
}
exports.getLocale = getLocale;
function formatCurrency(a, currency, locale, precision = 2) {
    if (!currency) {
        currency = 'GBP';
    }
    if (!locale) {
        locale = 'en-GB';
    }
    if (!a) {
        a = 0;
    }
    const amount = typeof a === 'string' ? Number.parseFloat(a) : a;
    currency = currency.toUpperCase();
    let value = format(amount, {
        code: currency,
        locale: locale,
        precision,
    }).replace(/\s/g, ' ');
    if (currency?.toLowerCase() === 'mxn') {
        value = value.replace('$', 'mex$');
    }
    return value;
}
exports.formatCurrency = formatCurrency;
function formatCurrencyNumber(a, locale) {
    const options = {
        minimumFractionDigits: 2,
        maximumFractionDigits: 2,
    };
    return Number(a).toLocaleString(locale, options);
}
exports.formatCurrencyNumber = formatCurrencyNumber;
